### PR TITLE
add AsyncEnumerable support to thread safe sockets

### PR DESCRIPTION
**Pull request by somdoron**

### CODE REVIEW

File: src/NetMQ/ReceiveThreadSafeSocketExtensions.cs [hunk 1]
```diff
@@ -117,6 +119,31 @@ public static bool TryReceiveBytes(this IThreadSafeInSocket socket, [NotNullWhen
...
+        
+        /// <summary>
+        /// Provides a consuming IAsyncEnumerable for receiving messages from the socket.
+        /// </summary>
+        /// <param name="socket">The socket to receive from.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>An IAsyncEnumerable that receive and returns messages from the socket.</returns>
+        /// <exception cref="System.OperationCanceledException">The token has had cancellation requested.</exception>
+        public static async IAsyncEnumerable<byte[]> ReceiveBytesAsyncEnumerable(
+            [NotNull] this IThreadSafeInSocket socket,
```
Comment (drewnoakes) [inline src/NetMQ/ReceiveThreadSafeSocketExtensions.cs:134]:
```suggestion
            this IThreadSafeInSocket socket,
```

In a nullable context (i.e. where `#nullable enable` is used) then `[NotNull]`/`[DisallowNull]` is the default for un-annotated references.

Note that if you want to annotate that null is not a valid input (eg. on an unconstrainted generic reference) you must use `DisallowNull` (or `MaybeNull` for the inverse). `NotNull` in this case implies the reference is not null _after_ the call returns, and is usually used for `ref`/`out` params.

Ditto three more times below.

File: src/NetMQ.Tests/ClientServer.cs [hunk 2]
```diff
@@ -71,7 +79,64 @@ public async void AsyncWithCancellationToken()
...
+            {
+                client.Send("0");
+            }
+            
+            // Send the end message to both of the threads
+            client.Send("1");
+            client.Send("1");
+
+            t1.Wait();
+            t2.Wait();
```
Comment (drewnoakes) [inline src/NetMQ.Tests/ClientServer.cs:137]: nit: If this test could deadlock then awaiting these tasks with a timeout would be good.

File: src/NetMQ.Tests/ClientServer.cs [hunk 3]
```diff
@@ -71,7 +79,64 @@ public async void AsyncWithCancellationToken()
...
+
+            for (int i = 0; i < 15000; i++)
+            {
+                client.Send("0");
+            }
+            
+            // Send the end message to both of the threads
+            client.Send("1");
+            client.Send("1");
+
```
Comment (drewnoakes) [inline src/NetMQ.Tests/ClientServer.cs:135]: Could have a `totalCount` and assert it's 15,002 here to verify no dropped messages.

File: src/NetMQ.Tests/NetMQ.Tests.csproj [hunk 4]
```diff
@@ -8,7 +8,7 @@
     <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
     <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">$(PackageTargetFallback);netcoreapp1.0;portable-net45+win8</PackageTargetFallback>
     <IsTestProject>true</IsTestProject>
-    <TargetFrameworks>netcoreapp2.1;net47</TargetFrameworks>
+    <TargetFrameworks>netcoreapp3.1;netcoreapp2.1;net47</TargetFrameworks>
```
Comment (drewnoakes) [inline src/NetMQ.Tests/NetMQ.Tests.csproj:11]: Consider dropping `netcoreapp2.1`.

Comment (somdoron) [reply src/NetMQ.Tests/NetMQ.Tests.csproj:11]: This is how I test netstandard2.0, so to drop netcoreapp2.1 would also mean to drop netstandard2.0

File: src/NetMQ/ReceiveThreadSafeSocketExtensions.cs (hunk 1 above)
Comment (somdoron) [reply src/NetMQ/ReceiveThreadSafeSocketExtensions.cs]: thanks, fixed.

**Comment by somdoron** [comment]
Added, thanks for the review, fixed it all.

**Pull request merged**